File: python/algorithms/binary_search.py
def binary_search(haystack: list[int], needle: int) -> int:
    """
    Performs binary search on a sorted array
    Returns index of needle if found, -1 otherwise
    """
    low, high = 0, len(haystack) - 1

    while low <= high:
        middle = (low + high) // 2
        value = haystack[middle]

        if value == needle:
            return True
        elif value < needle:
            low = middle + 1
        else:
            high = middle - 1

    return False

File: go/algorithms/search/binarysearch.go
// algorithms/search/binarysearch.go
package search

func BinarySearch(haystack []int, needle int) bool {
	low := 0
	high := len(haystack) - 1
	for low <= high {
		middle := (low + high) / 2
		if haystack[middle] == needle {
			return true
		} else if haystack[middle] < needle {
			low = middle + 1
		} else {
			high = middle - 1
		}
	}
	return false
}
